File: content/develop/use-cases/rate-limiter/nodejs/tokenBucket.js
/**
 * Token Bucket Rate Limiter
 *
 * A Redis-based token bucket rate limiter implementation using Lua scripts
 * for atomic operations.
 *
 * The token bucket algorithm allows requests at a controlled rate by maintaining
 * a bucket of tokens that refills over time. Each request consumes a token, and
 * requests are denied when the bucket is empty.
 *
 * @module tokenBucket
 */

const crypto = require("crypto");

// Lua script for atomic token bucket operations
const TOKEN_BUCKET_SCRIPT = `
local key = KEYS[1]
local capacity = tonumber(ARGV[1])
local refill_rate = tonumber(ARGV[2])
local refill_interval = tonumber(ARGV[3])
local now = tonumber(ARGV[4])

-- Get current state or initialize
local bucket = redis.call('HMGET', key, 'tokens', 'last_refill')
local tokens = tonumber(bucket[1])
local last_refill = tonumber(bucket[2])

-- Initialize if this is the first request
if tokens == nil then
    tokens = capacity
    last_refill = now
end

-- Calculate token refill
local time_passed = now - last_refill
local refills = math.floor(time_passed / refill_interval)

if refills > 0 then
    tokens = math.min(capacity, tokens + (refills * refill_rate))
    last_refill = last_refill + (refills * refill_interval)
end

-- Try to consume a token
local allowed = 0
if tokens >= 1 then
    tokens = tokens - 1
    allowed = 1
end

-- Update state
redis.call('HMSET', key, 'tokens', tokens, 'last_refill', last_refill)

-- Return result: allowed (1 or 0) and remaining tokens
return {allowed, tokens}
`;

/**
 * Token bucket rate limiter using Redis for distributed rate limiting.
 *
 * The token bucket maintains a fixed capacity of tokens that refill at a
 * constant rate. Each request consumes one token. When the bucket is empty,
 * requests are denied until tokens refill.
 *
 * @example
 * const { createClient } = require('redis');
 * const { TokenBucket } = require('./tokenBucket');
 *
 * const client = createClient();
 * await client.connect();
 *
 * const limiter = new TokenBucket({
 *   capacity: 10,
 *   refillRate: 1,
 *   refillInterval: 1.0,
 *   redisClient: client
 * });
 *
 * const { allowed, remaining } = await limiter.allow('user:123');
 * if (allowed) {
 *   console.log(`Request allowed. ${remaining} tokens remaining.`);
 * } else {
 *   console.log('Request denied. Rate limit exceeded.');
 * }
 */
class TokenBucket {
  /**
   * Create a new TokenBucket rate limiter.
   *
   * @param {Object} options - Configuration options.
   * @param {number} [options.capacity=10] - Maximum number of tokens in the bucket.
   * @param {number} [options.refillRate=1] - Number of tokens added per refill interval.
   * @param {number} [options.refillInterval=1.0] - Time in seconds between refills.
   * @param {import('redis').RedisClientType} [options.redisClient] - Redis client instance.
   */
  constructor({ capacity = 10, refillRate = 1, refillInterval = 1.0, redisClient } = {}) {
    this.capacity = capacity;
    this.refillRate = refillRate;
    this.refillInterval = refillInterval;
    this.redis = redisClient;

    // Calculate SHA1 of the script for EVALSHA
    this._scriptSha = crypto
      .createHash("sha1")
      .update(TOKEN_BUCKET_SCRIPT)
      .digest("hex");
    this._scriptLoaded = false;
  }

  /**
   * Ensure the Lua script is loaded into Redis.
   * @private
   */
  async _ensureScriptLoaded() {
    if (!this._scriptLoaded) {
      try {
        this._scriptSha = await this.redis.scriptLoad(TOKEN_BUCKET_SCRIPT);
        this._scriptLoaded = true;
      } catch {
        // If loading fails, we'll fall back to EVAL
      }
    }
  }

  /**
   * Check if a request should be allowed for the given key.
   *
   * @param {string} key - The rate limit key (e.g., 'user:123', 'api:endpoint:xyz').
   * @returns {Promise<{allowed: boolean, remaining: number}>} Result with allowed status and remaining tokens.
   *
   * @example
   * const { allowed, remaining } = await limiter.allow('user:123');
   * console.log(`Allowed: ${allowed}, Remaining: ${remaining}`);
   */
  async allow(key) {
    await this._ensureScriptLoaded();

    const now = Date.now() / 1000; // Current time in seconds

    let result;
    try {
      // Try EVALSHA first (faster if script is cached)
      result = await this.redis.evalSha(this._scriptSha, {
        keys: [key],
        arguments: [
          String(this.capacity),
          String(this.refillRate),
          String(this.refillInterval),
          String(now),
        ],
      });
    } catch (err) {
      if (err.message && err.message.includes("NOSCRIPT")) {
        // Script not in cache, use EVAL and reload
        result = await this.redis.eval(TOKEN_BUCKET_SCRIPT, {
          keys: [key],
          arguments: [
            String(this.capacity),
            String(this.refillRate),
            String(this.refillInterval),
            String(now),
          ],
        });
        this._scriptLoaded = false;
      } else {
        throw err;
      }
    }

    const allowed = Boolean(result[0]);
    const remaining = Number(result[1]);

    return { allowed, remaining };
  }
}

module.exports = { TokenBucket, TOKEN_BUCKET_SCRIPT };

File: content/develop/use-cases/rate-limiter/go/token_bucket.go
// Package ratelimiter provides a Redis-based token bucket rate limiter.
//
// The token bucket algorithm allows requests at a controlled rate by maintaining
// a bucket of tokens that refills over time. Each request consumes a token, and
// requests are denied when the bucket is empty.
//
// This implementation uses a Lua script executed atomically in Redis to ensure
// correct behavior under concurrent access.
package ratelimiter

import (
	"context"
	"crypto/sha1"
	"fmt"
	"time"

	"github.com/redis/go-redis/v9"
)

// tokenBucketScript is the canonical Lua script for atomic token bucket operations.
// All language implementations use this exact script for behavioral consistency.
const tokenBucketScript = `
local key = KEYS[1]
local capacity = tonumber(ARGV[1])
local refill_rate = tonumber(ARGV[2])
local refill_interval = tonumber(ARGV[3])
local now = tonumber(ARGV[4])

-- Get current state or initialize
local bucket = redis.call('HMGET', key, 'tokens', 'last_refill')
local tokens = tonumber(bucket[1])
local last_refill = tonumber(bucket[2])

-- Initialize if this is the first request
if tokens == nil then
    tokens = capacity
    last_refill = now
end

-- Calculate token refill
local time_passed = now - last_refill
local refills = math.floor(time_passed / refill_interval)

if refills > 0 then
    tokens = math.min(capacity, tokens + (refills * refill_rate))
    last_refill = last_refill + (refills * refill_interval)
end

-- Try to consume a token
local allowed = 0
if tokens >= 1 then
    tokens = tokens - 1
    allowed = 1
end

-- Update state
redis.call('HMSET', key, 'tokens', tokens, 'last_refill', last_refill)

-- Return result: allowed (1 or 0) and remaining tokens
return {allowed, tokens}
`

// TokenBucketConfig holds the configuration for a TokenBucket rate limiter.
type TokenBucketConfig struct {
	// Capacity is the maximum number of tokens in the bucket.
	// Default: 10
	Capacity int

	// RefillRate is the number of tokens added per refill interval.
	// Default: 1
	RefillRate float64

	// RefillInterval is the time between refills.
	// Default: 1 second
	RefillInterval time.Duration

	// Client is the Redis client to use. Required.
	Client *redis.Client
}

// TokenBucket is a Redis-based token bucket rate limiter.
//
// It maintains a fixed capacity of tokens that refill at a constant rate.
// Each request consumes one token. When the bucket is empty, requests are
// denied until tokens refill.
//
// Example:
//
//	rdb := redis.NewClient(&redis.Options{Addr: "localhost:6379"})
//	limiter := ratelimiter.NewTokenBucket(ratelimiter.TokenBucketConfig{
//	    Capacity:       10,
//	    RefillRate:     1,
//	    RefillInterval: time.Second,
//	    Client:         rdb,
//	})
//	allowed, remaining, err := limiter.Allow(ctx, "user:123")
//	if allowed {
//	    fmt.Printf("Request allowed. %.0f tokens remaining.\n", remaining)
//	} else {
//	    fmt.Println("Request denied. Rate limit exceeded.")
//	}
type TokenBucket struct {
	client         *redis.Client
	capacity       int
	refillRate     float64
	refillInterval float64 // in seconds
	scriptSHA      string
	scriptLoaded   bool
}

// NewTokenBucket creates a new TokenBucket rate limiter with the given configuration.
//
// Default values are applied for zero-value fields:
//   - Capacity: 10
//   - RefillRate: 1.0
//   - RefillInterval: 1 second
//
// The Client field is required and must not be nil.
func NewTokenBucket(cfg TokenBucketConfig) *TokenBucket {
	if cfg.Capacity == 0 {
		cfg.Capacity = 10
	}
	if cfg.RefillRate == 0 {
		cfg.RefillRate = 1.0
	}
	if cfg.RefillInterval == 0 {
		cfg.RefillInterval = time.Second
	}

	h := sha1.New()
	h.Write([]byte(tokenBucketScript))
	sha := fmt.Sprintf("%x", h.Sum(nil))

	return &TokenBucket{
		client:         cfg.Client,
		capacity:       cfg.Capacity,
		refillRate:     cfg.RefillRate,
		refillInterval: cfg.RefillInterval.Seconds(),
		scriptSHA:      sha,
	}
}

// ensureScriptLoaded loads the Lua script into Redis if it hasn't been loaded yet.
func (tb *TokenBucket) ensureScriptLoaded(ctx context.Context) {
	if !tb.scriptLoaded {
		sha, err := tb.client.ScriptLoad(ctx, tokenBucketScript).Result()
		if err == nil {
			tb.scriptSHA = sha
			tb.scriptLoaded = true
		}
	}
}

// Allow checks if a request should be allowed for the given key.
//
// It atomically checks and updates the token bucket state in Redis.
// Returns whether the request is allowed, the number of remaining tokens,
// and any error encountered.
//
// Example:
//
//	allowed, remaining, err := limiter.Allow(ctx, "user:123")
//	if err != nil {
//	    log.Printf("Rate limiter error: %v", err)
//	    // Fail open or closed depending on your policy
//	}
func (tb *TokenBucket) Allow(ctx context.Context, key string) (bool, float64, error) {
	tb.ensureScriptLoaded(ctx)

	now := float64(time.Now().UnixMicro()) / 1e6

	args := []interface{}{
		tb.capacity,
		tb.refillRate,
		tb.refillInterval,
		now,
	}

	// Try EVALSHA first (faster if script is cached)
	result, err := tb.client.EvalSha(ctx, tb.scriptSHA, []string{key}, args...).Int64Slice()
	if err != nil {
		// Script not in cache, fall back to EVAL
		result, err = tb.client.Eval(ctx, tokenBucketScript, []string{key}, args...).Int64Slice()
		if err != nil {
			return false, 0, fmt.Errorf("token bucket eval failed: %w", err)
		}
		tb.scriptLoaded = false
	}

	allowed := result[0] == 1
	remaining := float64(result[1])

	return allowed, remaining, nil
}
